### PR TITLE
RIO-51 feat(CollectionActor): Support PreUpdate Hook

### PR DESCRIPTION
**Pull request by HatemAlSum**

### CODE REVIEW

File: really-core/src/main/java/io/really/model/PreUpdate.java [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package io.really.model;
```
Comment (ashihaby) [inline really-core/src/main/java/io/really/model/PreUpdate.java]: Where is the license?

File: really-core/src/main/scala/io/really/model/CollectionActor.scala [hunk 2]
```diff
@@ -310,8 +310,9 @@ class CollectionActor(globals: ReallyGlobals) extends PersistentActor
       unstashAll()
       goto(WithModel) using ModelData(model)
 
-    case Event(ObjectExists(r), data @ ValidationReferences(request: Update, obj, model, expectedReferences, Nil, requester)) if expectedReferences.size == 1 =>
-      updateAndReply(request, obj, model, requester)
+    case Event(ObjectExists(r), data @ ValidationReferences(request: Update, after, model, expectedReferences, Nil, requester)) if expectedReferences.size == 1 =>
```
Comment (ashihaby) [inline really-core/src/main/scala/io/really/model/CollectionActor.scala]: long line

File: really-core/src/main/scala/io/really/model/CollectionActor.scala [hunk 3]
```diff
@@ -310,8 +310,9 @@ class CollectionActor(globals: ReallyGlobals) extends PersistentActor
...
+      //ToDo :  UpdateAndReplay Method need two jsObject (before , after)
```
Comment (ashihaby) [inline really-core/src/main/scala/io/really/model/CollectionActor.scala]: I din't get what you mean by that, and why you didn't do that?

File: really-core/src/main/scala/io/really/model/Model.scala [hunk 4]
```diff
@@ -148,7 +169,22 @@ package object model {
...
+      executePreUpdateValidator match {
+        case Some(preUpdateValidator: PreUpdate) =>
+          try {
+            preUpdateValidator.preUpdate(before.toString, after.toString, fieldsChanged.toArray)
+            ModelHookStatus.Succeeded
+          } catch {
+            case te: ModelHookStatus.JSValidationError => te.terminated
+            case se: ScriptException =>
+              //TODO Log the error
+              println("PreUpdate Script Execution Error: " + se)
```
Comment (ashihaby) [inline really-core/src/main/scala/io/really/model/Model.scala]: println?

File: really-core/src/main/scala/io/really/model/CollectionActor.scala [hunk 5]
```diff
@@ -310,8 +310,9 @@ class CollectionActor(globals: ReallyGlobals) extends PersistentActor
...
+      updateAndReply(request, after, after, model, requester)
```
Comment (ashihaby) [inline really-core/src/main/scala/io/really/model/CollectionActor.scala]: `before` and `after` are the same here, both of them takes `after` value

**Comment by ashihaby** [comment]
ok to test

**Comment by AhmedSoliman** [comment]
ok to test

**Comment by montaro** [comment]
retest this please

**Comment by montaro** [comment]
@HatemAlSum Tests are failing

**Pull request closed without merging**